synapse/handlers/register.py

```diff
@@ -448,9 +463,15 @@ async def _join_rooms(self, user_id: str) -> None:
                     )
 
                 # Calculate whether the room requires an invite or can be
-                # joined directly. Note that unless a join rule of public exists,
-                # it is treated as requiring an invite.
-                requires_invite = True
+                # joined directly. By default, we consider the room as requiring an
+                # invite if the homeserver is in the room (unless told otherwise by the
+                # join rules). Otherwise we consider it as being joinable, at the risk of
+                # failing to join, but in this case there's little more we can do since
+                # we don't have a local user in the room to craft up an invite with.
+                hosts_in_room = await self.state_handler.get_current_hosts_in_room(
+                    room_id
+                )
+                requires_invite = self.server_name in hosts_in_room
```

Comment:
I'd use `store.is_host_joined` to avoid have to do state res here (in general using store is better than state handler), plus its got a cache on it.

Reply:
Hmm, not sure how I've missed this method when I was looking for one to use. Thanks!